### PR TITLE
Add layer support

### LINKED ISSUE #112: Layers
Come up with a better way of thinking about the layering of entities instead of just render-priority on the drawables.

### PR DESCRIPTION
**Pull request by jimmcnulty41**
Resolves #112

### CODE REVIEW

File: src/duckling/canvas/_toolbars/top-toolbar.component.ts [hunk 1]
```diff
@@ -72,7 +82,9 @@ export class TopToolbarComponent {
     constructor(public store : StoreService,
                 public project : ProjectService,
                 public toolService : ToolService,
-                private _viewContainer : ViewContainerRef) {
+                private _viewContainer : ViewContainerRef,
```
Comment (jeffswenson) [inline src/duckling/canvas/_toolbars/top-toolbar.component.ts]: You aren't using the MdDialog imports or private fields.

File: src/duckling/canvas/drawing/entity-drawer.service.ts [hunk 2]
```diff
@@ -2,9 +2,10 @@ import {Component, Injectable} from '@angular/core';
 import {Container, DisplayObject} from 'pixi.js';
 
 import {BaseAttributeService} from '../../entitysystem/base-attribute.service';
-import {AssetService} from '../../project';
+import {AssetService, RequiredAssetService} from '../../project';
```
Comment (jeffswenson) [inline src/duckling/canvas/drawing/entity-drawer.service.ts]: You don't use RequiredAssetService.

File: src/duckling/canvas/map-editor.component.ts [hunk 3]
```diff
@@ -135,7 +137,8 @@ export class MapEditorComponent implements AfterViewInit, OnDestroy {
                 public projectService : ProjectService,
                 private _entityDrawerService : EntityDrawerService,
                 private _keyboardService : KeyboardService,
-                private _entityBoxService : EntityBoxService) {
+                private _entityBoxService : EntityBoxService,
```
Comment (jeffswenson) [inline src/duckling/canvas/map-editor.component.ts]: You aren't using _entityBoxService.

File: src/duckling/controls/layer-dialog.component.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+.show-hide-layer-button {
```
Comment (jeffswenson) [inline src/duckling/controls/layer-dialog.component.scss]: I don't think this needs to exist.

File: src/duckling/controls/layer-dialog.component.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import {
```
Comment (jeffswenson) [inline src/duckling/controls/layer-dialog.component.ts]: This file doesn't belong in controls/. It should be in the same folder as the layers service.

File: src/duckling/controls/layer-dialog.component.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        this._layerServiceSubscription.unsubscribe;
+    }
+
+    toggleLayerVisibility(layer : Layer) {
+        this._entityLayerService.toggleLayerVisibility(layer.layerName);
+    }
+
+    private _refreshLayers(){
+        this.layers = Array.from(this._entityLayerService.getLayers());
+        this.layers.sort((a, b) => {
```
Comment (jeffswenson) [inline src/duckling/controls/layer-dialog.component.ts]: This sorts 2 after 10.

File: src/duckling/entitysystem/services/entity-layer.service.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ */
+export type LayerGetter = (attribute : Attribute) => string;
+
+/**
+ * The EntityLayerService is used to retrieve the layer of an entity.
+ */
+@Injectable()
+export class EntityLayerService extends BaseAttributeService<LayerGetter> {
+
+    private _hiddenLayers : {[layerKey : string] : boolean} = {};
```
Comment (jeffswenson) [inline src/duckling/entitysystem/services/entity-layer.service.ts]: This should be stored in the redux store. That allows us to implement hot reloading in the future. You can also use that to make layer changes undo-able. Look at the selection service as an example.

File: src/duckling/entitysystem/services/entity-selection.service.ts [hunk 8]
```diff
@@ -37,6 +40,16 @@ export class EntitySelectionService {
         }
     }
 
+    private _getVisibleEntities(entities : Array<Entity>) : Array<Entity> {
```
Comment (jeffswenson) [inline src/duckling/entitysystem/services/entity-selection.service.ts]: This should be a method on the layer service.

**Review by jeffswenson — CHANGES_REQUESTED** [review]

File: src/duckling/canvas/_toolbars/top-toolbar.component.ts (hunk 1 above)
Comment (jimmcnulty41) [reply src/duckling/canvas/_toolbars/top-toolbar.component.ts]: Cool

File: src/duckling/canvas/drawing/entity-drawer.service.ts (hunk 2 above)
Comment (jimmcnulty41) [reply src/duckling/canvas/drawing/entity-drawer.service.ts]: Cool

File: src/duckling/canvas/map-editor.component.ts (hunk 3 above)
Comment (jimmcnulty41) [reply src/duckling/canvas/map-editor.component.ts]: Cool

File: src/duckling/controls/layer-dialog.component.scss (hunk 4 above)
Comment (jimmcnulty41) [reply src/duckling/controls/layer-dialog.component.scss]: Cool

File: src/duckling/controls/layer-dialog.component.ts (hunk 5 above)
Comment (jimmcnulty41) [inline src/duckling/controls/layer-dialog.component.ts]: Cool

File: src/duckling/entitysystem/services/entity-selection.service.ts (hunk 8 above)
Comment (jimmcnulty41) [reply src/duckling/entitysystem/services/entity-selection.service.ts]: Cool

File: src/duckling/entitysystem/services/entity-layer.service.ts (hunk 7 above)
Comment (tlein) [reply src/duckling/entitysystem/services/entity-layer.service.ts]: cool

File: src/duckling/controls/layer-dialog.component.ts (hunk 6 above)
Comment (tlein) [reply src/duckling/controls/layer-dialog.component.ts]: cool

**Pull request merged**